crowd_anki/utils/notifier.py

```diff
@@ -16,16 +17,27 @@ def warning(self, title: str, message: str):
     def error(self, title: str, message: str):
         pass
 
+def run_closure_in_main(closure):
+    """Run the closure in the main thread.
+
+This is necessary, because GUI operations in a background thread cause
+a crash.  For instance, export is now run in a background thread.
+
+    """
+    if aqt.mw.inMainThread():
+        closure()
+    else:
+        aqt.mw.taskman.run_on_main(lambda: aqt.mw.progress.timer(50, closure, False))
```

Comment:
any reason not to have 0 there?

Reply:
or, actually what does that number signify? (I initially assumed - delay)

Reply:
It's the [initial delay time](https://github.com/ankitects/anki/blob/8555a977cbf4b3633eaeed328619c796aaba12b0/qt/aqt/progress.py#L32). If I understand correctly, Anki will try to run the closure after the initial delay (in ms) and then try repeatedly in 100 ms intervals, until it's possible to do so.

I picked 50 ms as small enough not to result in any (significant) visible latency, but large enough that 

I didn't use 0, since:

1. Anki itself doesn't seem to use 0 anywhere where it uses `mw.progress.timer`.
2. I thought (without testing the timing explicitly) that since Anki can't run the closure immediately (since without the timer Anki hangs), it's better to wait for 50 ms before trying for the first time, as then we might have a chance of succeeding. (With `0`, we try to run the closure at times (0, 100, 200, 300...), but presumably the first try will fail, so we'll probably end up running at 100 ms. With `50`, we try at (50, 150, 250...), but there's a chance the first time will actually succeed, so we might end up running at 50 ms, earlier than with `0`.)

However, explicitly testing the timing (with the following ugly make-shift contraption:

<details>

```python3
import time
# ...
    def warning(self, title: str, message: str):
        print(time.time())
        tttt = time.time()
        def foo():
            print(time.time())
            print(time.time() - tttt)
            aqt.utils.showWarning(message, title=title)
        run_closure_in_main(foo)
```

</details>

) suggests that using `0` is still faster. (For `50`, the times are roughly in 70-90 ms range, while for `0` they're in the 10-50 ms range. For `10`, they're in the 30-70 ms range.)

A timeout of 0 is explicitly considered in the [official Qt docs](https://doc.qt.io/qt-5/qtimer.html#details) valid, so it's not a weird edge-case that just accidentally works, so I'll switch to it!

Reply:
Thanks for diving deeper on this!